File: Valid Anagram/validAnagram.java
/**
 https://leetcode.com/problems/valid-anagram/
time complexity: O(n) where n is the max of length of s and t strings
space complexity: O(1)
approach: if lengths of s and t string are not same, return false
maintain a frequency array
iterate over the s string and update the frequency array
iterate over the t string and decrease the frequencies of the characters
if any of the frequency of the frequency array is not 0, return false
 */
class Solution {
    public boolean isAnagram(String s, String t) {
        int sLen = s.length();
        int tLen = t.length();
        if(sLen != tLen) return false;
        int[] sFreq = new int[256];
        for(char c : s.toCharArray())
        {
            sFreq[c]++;
        }
        for(char c : t.toCharArray())
        {
            sFreq[c]--;
        }
        for(int i : sFreq)
        {
            if(i != 0) return false;
        }
        return true;
    }
}

File: Valid Anagram/valid_anagram.py
"""
https://leetcode.com/problems/valid-anagram/
Time complexity: O(n) where n is the maximum length of strings s and t.

Space complexity: O(1) as the size of the frequency array 
                     (256 for ASCII characters) is constant.
Approach:
    1. If the lengths of strings s and t are different, they cannot be anagrams.
    2. Create a frequency array to store the counts of each character.
    3. Iterate through string s and increment the frequency of each character in the array.
    4. Iterate through string t and decrement the frequency of each character in the array.
    5. If any frequency in the array is not zero, the strings are not anagrams.
"""
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        frequency = [0] * 256
        for c in s:
            frequency[ord(c)] = frequency[ord(c)] + 1
        for c in t:
            frequency[ord(c)] = frequency[ord(c)] - 1
        for i in frequency:
            if i != 0:
                return False
        return True
        